### PR TITLE
Moved creating SiPixelTemplateStore from DB to an ESProducer

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

All DB construction was switched to using EventSetup. Cases using reading from files were kept local to module.
The new ESProducer was added as a Task to all needed Sequences.

#### PR validation:

Code compiles. All changed python files are readable by python3. No cmsRun jobs were run.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
please test

File: CondFormats/SiPixelTransient/src/T_EventSetup_SiPixelTemplateStore.cc [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Original Author:  Christopher Jones
+//         Created:  Tue, 08 Aug 2023 15:33:48 GMT
+//
+
+// system include files
+
+// user include files
+#include "CondFormats/SiPixelTransient/interface/SiPixelTemplate.h"
+#include "FWCore/Framework/interface/eventsetuprecord_registration_macro.h"
```
Comment (makortel) [inline CondFormats/SiPixelTransient/src/T_EventSetup_SiPixelTemplateStore.cc:17]: I believe this should be
```suggestion
#include "FWCore/Utilities/interface/typelookup.h"
```
which would then avoid adding the dependence on `FWCore/Framework` to this package.

**Comment by Dr15Jones** [comment]
please test

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py [hunk 2]
```diff
@@ -65,5 +66,6 @@
     ALCARECOCalCosmicsFilterForSiPixelLorentzAngleMCS *
     ALCARECOPixelLATrackFilterRefitMCS *
     ALCARECOSiPixelLACalibMCS *
-    MEtoEDMConvertSiPixelLorentzAngleMCS 
+    MEtoEDMConvertSiPixelLorentzAngleMCS,
+    cms.Task(SiPixelTemplateStoreESProducer)
```
Comment (tvami) [inline Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:70]: for my education, why is this line with `cms.Task()` while none of the above are? Thanks!

Comment (mmusich) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:70]: https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideAboutPythonConfigFile#Behavior_when_an_ESProducer_ESSo

**Comment by Dr15Jones** [comment]
please test

File: Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py (hunk 2 above)
Comment (tvami) [reply Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiPixelLorentzAngleMCS_cff.py:70]: I see, thanks Marco and Chris!

**Comment by Dr15Jones** [comment]
@Martin-Grunewald how does one go about updating the HLT configs in CMSSW?

**Comment by Martin-Grunewald** [comment]
@Dr15Jones 
Please add the process.load of the cfi for the ESProducer to the customisation file
https://github.com/cms-sw/cmssw/blob/master/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
is, add a customisation function ```customizeHLTfor42514``` there, and call it (at the bottom).
See the file for examples....
This is to get this PR through. Eventually we (TSG-STORM) will put the new module into ConfDb menus, redo the menu dumps, and remove the customisation code.

**Comment by Martin-Grunewald** [comment]
It is 42514 (not 42510)...

File: HLTrigger/Configuration/python/customizeHLTforCMSSW.py [hunk 3]
```diff
@@ -268,6 +268,11 @@ def customizeHLTfor42410(process):
 
     return process
 
+def customizeHLTfor42510(process):
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/customizeHLTforCMSSW.py:271]: 42514

**Comment by Dr15Jones** [comment]
please test

**Comment by Martin-Grunewald** [comment]
Looks like these are jobs running on rather old data/MC - eg, the devolpment HLT menu addOnTests (GRun etc.) now seem to run fine.

**Comment by mmusich** [comment]
> Looks like these jobs failing are running on rather old data/MC - eg, the devolpment HLT menu addOnTests (GRun etc.) now seem to run fine

well, e.g. for [this test](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-129713/34210/addOnTests/hlt_mc_Fake2/), the failed job is consuming: `auto:run2_mc_Fake2` which resolves to:

```
run2_mc_Fake2 : ('131X_mcRun2_asymptotic_v2', 'L1Menu_Collisions2016_v9_m2_xml,L1TUtmTriggerMenuRcd,frontier://FrontierProd/CMS_CONDITIONS,,2016-10-06 19:36:53.000')
```

(notice time of the snapshot: `2016-10-06 19:36:53.000`) but

```
 conddb list 131X_mcRun2_asymptotic_v2 | grep SiPixelTemplateDBObjectRcd
[2023-08-10 19:08:34,857] INFO: Connecting to pro [frontier://PromptProd/cms_conditions]
SiPixelTemplateDBObjectRcd                              -                                                              SiPixelTemplateDBObject_38T_2016_ultralegacymc_v2                          
SiPixelTemplateDBObjectRcd                              0T                                                             SiPixelTemplateDBObject_0T_v3_mc      
```

and in particular the creation time of [SiPixelTemplateDBObject_38T_2016_ultralegacymc_v2](https://cms-conddb.cern.ch/cmsDbBrowser/list/Prod/tags/SiPixelTemplateDBObject_38T_2016_ultralegacymc_v2) is `2019-07-16 21:59:27 (UTC)` (3 years later than the symbolic GT snapshot, no wonder that `No data of type "SiPixelTemplateDBObject" with label "" in record "SiPixelTemplateDBObjectESProducerRcd"`).
I am wondering though why before the PR the job was not failing already.

**Comment by makortel** [comment]
> > Looks like these jobs failing are running on rather old data/MC - eg, the devolpment HLT menu addOnTests (GRun etc.) now seem to run fine
> 
> well, e.g. for [this test](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-129713/34210/addOnTests/hlt_mc_Fake2/), the failed job is consuming: `auto:run2_mc_Fake2` which resolves to:
> ...
> I am wondering though why before the PR the job was not failing already.

The `EventSetupRecordDataGetter` doesn't seem to throw an exception if an ESProduct is missing.  The `SiPixelTemplateStoreESProducer` added in this PR requires (currently)  the `SiPixelTemplateDBObject` ESProduct to exist. I.e. the `EventSetupRecordDataGetter` consuming `std::vector<SiPixelTemplateStore>` causes `SiPixelTemplateStoreESProducer::produce()` to be run, which then throws the exception for a missing data product.

I'd guess no other module in the HLT addOnTests consumes the pixel templates, and therefore the jobs have succeeded before.

**Comment by mmusich** [comment]
> I'd guess no other module in the HLT addOnTests consumes the pixel templates, and therefore the jobs have succeeded before.

That's the part that puzzled me.  I thought that kind of test did run also RECO...

**Comment by makortel** [comment]
> > I'd guess no other module in the HLT addOnTests consumes the pixel templates, and therefore the jobs have succeeded before.
> 
> That's the part that puzzled me. I thought that kind of test did run also RECO...

Ah, good point, the test indeed appears to run RECO. I guess that would require a deeper investigation of what exactly is going on the job.

**Comment by missirol** [comment]
Fwiw, [1] worked for me using [2] on top of cb5704aa85b.

[1]
```
addOnTests.py && runTheMatrix.py -l 4.22,8.0,140.53,134.813,136.733,140.52,1005.0,1004.0,1052.0 --ibeos
```

[2]
```diff
diff --git a/Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py b/Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py
index 9dde4ff8f48..203a3b6641a 100644
--- a/Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py
+++ b/Calibration/TkAlCaRecoProducers/python/ALCARECOPromptCalibProdSiStripGainsAAG_cff.py
@@ -64,9 +64,11 @@ ALCARECOCalibrationTracksRefitAAG = TrackRefitter.clone(src = cms.InputTag("ALCA
 
 # refit and BS can be dropped if done together with RECO.
 # track filter can be moved in acalreco if no otehr users
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
 ALCARECOTrackFilterRefitAAG = cms.Sequence(ALCARECOCalibrationTracksAAG +
                                            offlineBeamSpot +
-                                           ALCARECOCalibrationTracksRefitAAG )
+                                           ALCARECOCalibrationTracksRefitAAG,
+                                           cms.Task(SiPixelTemplateStoreESProducer))
 
 # ------------------------------------------------------------------------------
 # This is the module actually doing the calibration
diff --git a/HLTrigger/Configuration/python/customizeHLTforCMSSW.py b/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
index 66eac258ddf..7c48a2d45c5 100644
--- a/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
+++ b/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
@@ -269,10 +269,12 @@ def customizeHLTfor42410(process):
     return process
 
 def customizeHLTfor42514(process):
-    process.load('RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi')
+    for foo in esproducers_by_type(process, 'SiPixelTemplateDBObjectESProducer'):
+        process.load('RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi')
+        break
 
     return process
-    
+
 # CMSSW version specific customizations
 def customizeHLTforCMSSW(process, menuType="GRun"):
 
@@ -287,5 +289,5 @@ def customizeHLTforCMSSW(process, menuType="GRun"):
     process = customizeHLTfor41632(process)
     process = customizeHLTfor42410(process)
     process = customizeHLTfor42514(process)
-    
+
     return process
diff --git a/RecoHI/HiTracking/python/HighPtTracking_PbPb_cff.py b/RecoHI/HiTracking/python/HighPtTracking_PbPb_cff.py
index 88920af36a6..a499ad0f192 100644
--- a/RecoHI/HiTracking/python/HighPtTracking_PbPb_cff.py
+++ b/RecoHI/HiTracking/python/HighPtTracking_PbPb_cff.py
@@ -4,6 +4,7 @@ import FWCore.ParameterSet.Config as cms
 from TrackingTools.TransientTrack.TransientTrackBuilder_cfi import * #also includes global tracking geometry
 from RecoTracker.TransientTrackingRecHit.TTRHBuilders_cff import *
 from RecoLocalTracker.SiPixelRecHits.PixelCPEESProducers_cff import * #cluster parameter estimator producer
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
 from RecoTracker.TkSeedingLayers.PixelLayerTriplets_cfi import *
 
 ### pixel primary vertices
@@ -31,4 +32,5 @@ hiBasicTrackingTask = cms.Task(hiPixelVerticesTask
                                 , hiPrimTrackCandidates
                                 , hiGlobalPrimTracks
                                 , hiTracksWithQualityTask
+                                , SiPixelTemplateStoreESProducer
                                 )
diff --git a/RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py b/RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py
index 978cb0b9d24..ffac02aab37 100644
--- a/RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py
+++ b/RecoTracker/SpecialSeedGenerators/python/cosmicDC_cff.py
@@ -51,6 +51,8 @@ cosmicDCTracks = RecoTracker.TrackProducer.CTFFinalFitWithMaterialP5_cff.ctfWith
     src = 'cosmicDCCkfTrackCandidates',
 )
 
+from RecoLocalTracker.SiPixelRecHits.SiPixelTemplateStoreESProducer_cfi import SiPixelTemplateStoreESProducer
+
 # Final Sequence
-cosmicDCTracksSeqTask = cms.Task( cosmicDCSeeds , cosmicDCCkfTrackCandidates , cosmicDCTracks )
+cosmicDCTracksSeqTask = cms.Task( cosmicDCSeeds , cosmicDCCkfTrackCandidates , cosmicDCTracks , SiPixelTemplateStoreESProducer )
 cosmicDCTracksSeq = cms.Sequence(cosmicDCTracksSeqTask)
```

**Comment by mmusich** [comment]
test parameters:
   * workflows = 1001.2, 1001.3, 1002.3, 1002.4, 1004, 1005, 1052, 1052.1

**Comment by Dr15Jones** [comment]
@missirol Thanks!

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
Please test

**Comment by Dr15Jones** [comment]
I am unable to run the 'unit tests' in Alignment/OfflineValidation as the files needed are inaccessible from FNAL, cmsdev and lxplus8. One such file is 
/eos/cms/store/user/cmsbuild/store/mc/RunIIWinter19PFCalibDR/Single_Pion_gun_E_2to200_13TeV_pythia8/GEN-SIM-RECO/2018ConditionsNoPU_105X_upgrade2018_realistic_v4-v1/270000/2BCBEBAB-BA94-B74F-9BFA-04F849A32558.root

it is VERY BAD to create unit tests that cannot be run at all possible sites.

**Comment by mmusich** [comment]
> it is VERY BAD to create unit tests that cannot be run at all possible sites.

It is VERY EASY to overcome that issue: `scram b runtests use-ibeos`

**Comment by Dr15Jones** [comment]
Doing `scram b runtests_DMRall use-ibeos` from FNAL fails as it seems to break frontier access.

**Comment by mmusich** [comment]
> from FNAL fails as it seems to break frontier access.

From any machine at CERN, it will work fine

**Comment by Dr15Jones** [comment]
Please test

**Comment by tvami** [comment]
+alca
 * tests pass, diffs in MsgLogger only

**Comment by missirol** [comment]
+hlt

(Also discussed in #40544 and #42504.)

**Comment by tvami** [comment]
+db

**Comment by missirol** [comment]
@cms-sw/fastsim-l2 , could you please review this PR ?

There are upcoming HLT PRs that require this PR to be merged first (changes to `customizeHLTforCMSSW.py`, which is touched by this PR).

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by mmusich** [comment]
Is it worth backporting this (at least to 13.2.X)?

**Comment by mmusich** [comment]
> Let backport only what is necessary,

AFAIK this makes a sizeable performance improvement in DQM memory requirement, so that's why I am asking if we want to make use of this or not (e.g. for the HI data-taking or for a second take on the pp 2023 reprocessing?)

**Comment by makortel** [comment]
> Is it worth backporting this (at least to 13.2.X)?

@cms-sw/ppd-l2 asked in the O&C today to backport this PR (+ https://github.com/cms-sw/cmssw/pull/42597) down to 13_0_X.

**Comment by jordan-martins** [comment]
thanks @makortel ! Indeed, this will come in hand with the most likely rereco of the 2023 DATA to happen in 130X, in case it happens

**Pull request merged**